doc/commands.rst

```diff
@@ -418,6 +418,46 @@ transformed into ``?x ≡ bool``.
 *WARNING* This feature is experimental and there is no sanity check
 performed on the rules.
 
+.. _coercion:
+
+``coercion``
+------------
+
+Lambdapi can be instructed to insert functions into terms whenever needed for
+typability. These functions are called *coercions*. For instance, assuming
+we have a type ``Float``, a type ``Int`` and a function
+``FloatOfInt : Int → Float``, the latter function can be declared
+as a coercion from integers to floats with the declaration
+
+::
+
+    coercion coerce Int Float $x ↪ FloatOfInt $x;
+
+Symbol ``coerce`` is a meta-function that computes the coercion.
+Whenever a term ``t`` of type ``Int`` is found when Lambadpi expected a
+``Float``, ``t`` will be replaced by ``coerce Int Float t`` and reduced.
+The declared coercion will allow the latter term to be reduced to
+``FloatOfInt t``.
+
+Coercions can call the function ``coerce`` recursively,
+which allows to write, e.g.
+
+::
+
+    coercion coerce (List $a) (List $b) $l ↪
+      map (λ e: El $a, coerce $a $b e) $l
+
+where ``Set: TYPE;``, ``List : Set → TYPE``, ``El : Set → TYPE`` and ``map`` is
+the usual mapping operator on lists such that ``map f (cons x l) = cons (f x)
+(map l)``.
+
+*WARNING* Coercions are still experimental, and will not mix well is
+metavariables. Indeed, the term ``coerce ?1 Float t`` will not reduce to
+``FloatOfInt t`` even if the equation ``?1 = Int`` has been registered during
+typing. Furthermore, it is unsafe to use coercions that compute to symbol that
+can be reduced to protected symbols, for the reduction may generate the
+protected symbols.
```

Comment:
Is it possible to have protected symbols? Isn't it the case that protected symbols are forbidden at scoping time?

Reply:
Sorry but I don't understand. This is your comment that is talking about protected symbols. So why are you talking about ghost symbols here. I am confused. What do you want to tell with your remark "it is unsafe to use coercions that compute to symbol that can be reduced to protected symbols"? Would you have an example? pair can reduce to the protected symbol pair'. Do you mean that we shouldn't use pair in a coercion rule? I guess no. So could you please clarify all this?

Reply:
I mentioned *ghost* symbols because they are the ones that are forbidden at scoping time, that's why I suspected a confusion. In that comment, I say that if a coercion rule contains a symbol that can reduce to a protected variant, then the protection can be violated.

For instance, let's take the `pair`, `pairp` protection scheme. In presence of the rule `coerce_rule coerce (El $a) (El (psub $a $p)) $e --> pair $a $p $e _`,
everything seems ok because we reduce to `pair`. However, in the function `reduce_coercions` of `infer.ml`, many subterms are reduced to `whnf`, so the reduction `pair $a $p $e _ --> pairp $a $p $e` may be triggered during the coercion elaboration.

Reply:
It seems related to some discussion we had before about limiting applicable rules when reducing a coercion. Is it right? Do you plan to do something about that?
Anyway isn't it problematic to say that this is unsafe but still have to use it in some developments? Perhaps you could instead say that the *current* implementation may be unsafe but we will try to do something about it.

Reply:
> It seems related to some discussion we had before about limiting applicable rules when reducing a coercion. Is it right?

It is, indeed.

> Anyway isn't it problematic to say that this is unsafe but still have to use it in some developments?

It is, but I don't use that version of lambdapi for PVS related things, I use another branch with more reduction options for symbols (I have added a tag for this).

> Perhaps you could instead say that the current implementation may be unsafe but we will try to do something about it.

Sounds good to me.

> Do you plan to do something about that?

I can't pledge to fix it, but I may be able to try something sometime.

Reply:
Aren't you going to include those "reduction options for symbols" in master?

Reply:
What are they?

Reply:
The mechanism I used is not very elegant, I added the tag `redopaque` for the symbol `pair`, and the reduction for coercion would reduce all but `redopaque` symbols.

Reply:
I can at least post an issue on the bugtracker.

Reply:
Can't we imagine to mark every symbol having a rule RHS containing a protected symbol as "progen" (add field sym_progen : bool ref) and do not reduce those symbols when reducing coercions?

Reply:
Yes, it seems it's what I implemented. It feels clunky, but it works.

Reply:
You prefer to add this in a separate PR?

Reply:
Yes I'd prefer, this one allows already interesting features (such as setting `El` as a coercion).